config.ru

```diff
@@ -4,4 +4,32 @@
 
 require_relative 'config/environment'
 
-run Rails.application
+ENV['PROMETHEUS_ENABLED'] ||= 'false'
+
+if ENV['PROMETHEUS_ENABLED'] == 'true'
+  require 'prometheus/middleware/collector'
+  require 'prometheus/middleware/exporter'
+
+  rackapp = Rack::Builder.app do
+    # rubocop:disable Metrics/LineLength
+    use Prometheus::Middleware::Collector
+    if ENV['PROMETHEUS_BASIC_AUTH_USERNAME'].present?
```

Comment:
I think we should wrap the usage of `Rack::Auth::Basic` into a custom conditional middleware. Something like:

```
config.middleware.use Prometheus::Middleware::Collector
class ConditionalAuthBasic
  def initialize(app)
    @app = app
  end

  def call(env)
    if env['REQUEST_PATH'].starts_with?("/metrics")
      basic_auth = Rack::Auth::Basic.new(@app, 'Ciao Prometheus Metrics') do |username, password|
          Rack::Utils.secure_compare(ENV['PROMETHEUS_BASIC_AUTH_USERNAME'], username) &&
            Rack::Utils.secure_compare(ENV['PROMETHEUS_BASIC_AUTH_PASSWORD'], password)
        end
      env = basic_auth.call(env)
    end

    @app.call(env)
  end
end
```

Then in `config/application.rb`. we only need to:

```
config.middleware.use ConditionalAuthBasic
config.middleware.use Prometheus::Middleware::Exporter
```

Reply:
Will be changed.

Reply:
Current approach is more readable.

Reply:
The approach that I mentioned is following Rack middleware conventions (chain middlewares instead of composite it). I think what you are doing is anti-pattern/convention.

Also you are nesting conditions with blocks that makes it harder to read. Based on these 2 arguments, I think my approach is more readable for any ruby developer 😅and SOLID

Reply:
@tareksamni Don't wanna start a discussion on which is more readable - sorry for that. This is more a subjective thing.

Could you push you enhancements (Rack middleware conventions: chain middlewares instead of composite it to this branch so we can merge it?

Reply:
I won't be able to change this any time soon. So feel free to go forward with this implementation till we get a better chance to change it. Also, would love to see this in production ASAP :)